### PR TITLE
Add multiple choice polls

### PR DESCRIPTION
**Pull request by owi92**
This PR adds an option to set the `maxSelections` parameter in the `createPollDialog`, and allows users to vote for more than one option based on that parameter. Instead of radio buttons, multiple choice polls feature checkboxes, which do however use the same styles as the radio buttons. If `maxSelections` is greater than 1, an indication telling the user the number of remaining votes is shown. Options can be deselected to vote for another option.
Closes [#23319](https://github.com/vector-im/element-web/issues/23319)

https://user-images.githubusercontent.com/94838646/201692244-229094f0-3427-4cfb-b6ed-e3fc191c15d3.mov

Signed-off-by: Ole Wieners olewieners@yahoo.com

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add multiple choice polls ([\#9519](https://github.com/matrix-org/matrix-react-sdk/pull/9519)). Contributed by @owi92.

### CODE REVIEW

File: res/css/views/elements/_StyledPollCheckbox.pcss
Comment (owi92) [inline res/css/views/elements/_StyledPollCheckbox.pcss:23]: These are the same styles used by the radio button, so the checkbox could also use the classes from the `_StyledRadioButton.pcss` file which would need a single additional line to include the `input[type="checkbox"] selector (see screenshot). I'm not sure if that would impede that file's reusability or break the component's independence, so I chose to copy the styles into a new file.
Bildschirmfoto 2022-11-14 um 16 15 00

**Comment by weeman1337** [comment]
They needs at least a review from design. Probably also from product. When merging this feature for Web it would not be available on the mobile apps. This could lead to a not so nice experience.

File: res/css/views/elements/_StyledPollCheckbox.pcss
Comment (weeman1337) [reply res/css/views/elements/_StyledPollCheckbox.pcss:23]: I would merge the styles.

File: src/components/views/elements/StyledPollCheckbox.tsx
Comment (weeman1337) [inline src/components/views/elements/StyledPollCheckbox.tsx:1]: Can't we use / extend https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/views/elements/StyledCheckbox.tsx for the use in polls?

File: src/components/views/messages/MPollBody.tsx
Comment (weeman1337) [inline src/components/views/messages/MPollBody.tsx:345]: To be safe I would suggest

```suggestion
            if (myVotes.length >= poll.maxSelections) {
```

File: src/i18n/strings/en_EN.json
Comment (weeman1337) [inline src/i18n/strings/en_EN.json:2533]: This is something that design should decide. I personally would expect something like „Number of votes per person“. It is „vote“ everywhere and not „selection“.

**Review by weeman1337 — CHANGES_REQUESTED** [review]
I think this is a good start. My main point in the code is the `StyledPollCheckbox` where there is already a `StyledCheckbox` available and poll uses the `StyledRadioButton`.

Also found an issue when clicking around: If you have an empty option you can select more available votes than options available.

image

image

Also I think we should cover multi vote polls with Cypress end-2-end tests.

**Comment by owi92** [comment]
Thank you for the review and suggestions! I believe I have addressed mosty everything with my lates push. I'm still missing the cypress end2end tests but will try to get them done tomorrow.

**Comment by pree** [comment]
Is there something missing for this to get merged?

**Comment by daniellekirkwood** [comment]
@jakewb-b It would be great to get your thoughts on this PR as an Enhancement to a feature you're close to. If this is something we want to move ahead with, let's engage with Design (@rufuskahler)

**Comment by rufuskahler** [comment]
@gaelledel – Please advise on the preferred design solution here.

**Comment by ggPeti** [comment]
+1 would love to have this feature, a team I'm involved in would like to conduct multi-select polls.
Would this only be available in the web client if merged?

**Comment by dbkr** [comment]
I've been looking at this to try & get it moving forward. I think the main obstacle at the moment is the question of what other clients do in the face of a multiple choice poll, since we'd be launching this feature on element web but not mobile. eg:
 *  If on mobile you don't even notice that you could have more votes, that might be ok
 * If you vote lots on web and then view the poll on mobile and it's not accurate, thats a problem

It will also be quite weird if you see other people clearly voting for more than one option, but can't do so yourself, although maybe not the end of the world.

I know other clients have support for this, but it being different in Element on different platforms is more of a problem.

In general, I would advise talking to us before implementing large features like this: they're not trivial to ship, unfortunately.

**Comment by dbkr** [comment]
I'm going to close this for now. If someone if able to pick up to work of establishing what the experience is, exactly, of having it on web and mobile, we can re-open and continue work on it.

**Pull request closed without merging**
